pkg/process/events/store.go

```diff
@@ -0,0 +1,277 @@
+	expiredBuffer *atomic.Int64 // how many events have been dropped due to a full buffer
+}
+
+// readPositiveInt reads a config stored in the given key and asserts that it's a positive value
+func readPositiveInt(key string) (int, error) {
+	i := ddconfig.Datadog.GetInt(key)
+	if i <= 0 {
+		return 0, errors.New("invalid setting. " + key + " should be > 0")
+	}
+
+	return i, nil
+}
+
+// NewRingStore creates a new RingStore to store process events
+func NewRingStore(client statsd.ClientInterface) (Store, error) {
+	maxItems, err := readPositiveInt("process_config.event_collection.store.max_items")
+	if err != nil {
+		return nil, err
+	}
+
+	maxPushes, err := readPositiveInt("process_config.event_collection.store.max_pending_pushes")
+	if err != nil {
+		return nil, err
+	}
+
+	maxPulls, err := readPositiveInt("process_config.event_collection.store.max_pending_pulls")
+	if err != nil {
+		return nil, err
+	}
+
+	statsInterval, err := readPositiveInt("process_config.event_collection.store.stats_interval")
+	if err != nil {
+		return nil, err
+	}
+
+	return &RingStore{
+		buffer:        make([]ringNode, maxItems),
+		head:          0,
+		tail:          0,
+		maxItems:      maxItems,
+		pushReq:       make(chan *pushRequest, maxPushes),
+		pullReq:       make(chan *pullRequest, maxPulls),
+		exit:          make(chan struct{}),
+		statsdClient:  client,
+		statsInterval: time.Duration(statsInterval) * time.Second,
+		expiredInput:  atomic.NewInt64(0),
+		expiredBuffer: atomic.NewInt64(0),
+	}, nil
+}
+
+// Run starts the RingStore. A go routine is created to serve push and pull requests in order to protect the underlying
+// storage from concurrent access
+func (s *RingStore) Run() {
+	log.Info("Starting the RingStore")
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.run()
+	}()
+}
+
+// Stop stops the RingStore's internal routine
+func (s *RingStore) Stop() {
+	log.Info("Stopping the RingStore")
+	close(s.exit)
+	s.wg.Wait()
+	log.Info("RingStore stopped")
+}
+
+// Push adds an event to the RingStore. If the store is full, the oldest event is dropped to make space for the new one
+// The done channel is optional. It's used to signal if the event has been successfully written to the Store
+func (s *RingStore) Push(e *model.ProcessEvent, done chan bool) {
+	r := &pushRequest{
+		event: e,
+		done:  done,
+	}
+
+	select {
+	case s.pushReq <- r:
+	default:
+		log.Trace("RingStore input channel is full, dropping event")
+		s.expiredInput.Inc()
+
+		if done != nil {
+			done <- false
+		}
+	}
+}
+
+// Pull returns all events stored in the RingStore
+func (s *RingStore) Pull(ctx context.Context, timeout time.Duration) ([]*model.ProcessEvent, error) {
+	timer := time.NewTimer(timeout)
+
+	q := &pullRequest{
+		results: make(chan []*model.ProcessEvent),
+	}
+
+	select {
+	case s.pullReq <- q:
+	default:
+		log.Warn("Can't Pull RingStore: too many pending requests")
+		return nil, errors.New("too many pending pull requests")
```

Comment:
We don't seem to be asserting on this error in tests?

Reply:
I've added a new `TestRingStorePullErrors` to test scenarios where pulls either time out or get rejected due to too many pending requests: https://github.com/DataDog/datadog-agent/pull/12159/commits/d476dc473670c15fae43b7c423f358503715eb7b.